File: cpp/space-age/space_age.cpp
#include "space_age.h"

namespace space_age
{
    double space_age::getAge(double param) const
    {
        return on_earth() / param;
    }
    long int space_age::seconds() const
    {
        return secs;
    }
    double space_age::on_earth() const
    {
        return secs / (double)31557600;
    }
    double space_age::on_mercury() const
    {
        return getAge(mercury);
    }
    double space_age::on_venus() const
    {
        return getAge(venus);
    }
    double space_age::on_mars() const
    {
        return getAge(mars);
    }
    double space_age::on_jupiter() const
    {
        return getAge(jupiter);
    }
    double space_age::on_saturn() const
    {
        return getAge(saturn);
    }
    double space_age::on_uranus() const
    {
        return getAge(uranus);
    }
    double space_age::on_neptune() const
    {
        return getAge(neptune);
    }

} // namespace space_age

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet (..), ageOn) where

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

ageOn :: Planet -> Float -> Float
ageOn Mercury seconds = inEarthYears seconds / 0.2408467
ageOn Venus seconds = inEarthYears seconds / 0.61519726
ageOn Earth seconds = inEarthYears seconds
ageOn Mars seconds = inEarthYears seconds / 1.8808158
ageOn Jupiter seconds = inEarthYears seconds / 11.862615
ageOn Saturn seconds = inEarthYears seconds / 29.447498
ageOn Uranus seconds = inEarthYears seconds / 84.016846
ageOn Neptune seconds = inEarthYears seconds / 164.79132

inEarthYears :: Float -> Float
inEarthYears seconds = seconds / 31557600
